### PR TITLE
Improve testing framework

### LINKED ISSUE #13: Improved logging in the testing framework
# Description 
The testing framework should offer the ability to log the line number where the assert failed.  This can be achieved through the usage of the ``__LINE__``, and will greatly improve testing, as currently the test can only output if it failed or succeeded.  This will involve a significant change to the way that the testing process works.

# Requirements

- [x] Return the line number on failed asserts
- [x] ~Create a preprocessor definition that turns this off~

### PR DESCRIPTION
**Pull request by garfr**
Improves on the existing bare-bones testing framework by adding line number logging and an ability to disable it.

### CODE REVIEW

File: utils/test/test_framework.h [hunk 1]
```diff
@@ -1,9 +1,35 @@
...
 #define SUCCESS 0
 #define FAIL -1
-#define ASSERT(expr) if(!(expr)) return FAIL
-#define ASSERT_IS_NONE(var) if(var != NULL) return FAIL
+
+#ifdef TEST_NO_PRINT_
+
+#define ASSERT(expr)                                                           \
+    if (!(expr))                                                               \
+        return FAIL;
```
Comment (bama4) [inline utils/test/test_framework.h]: No semicolon should be in the macro https://wiki.sei.cmu.edu/confluence/display/c/PRE11-C.+Do+not+conclude+macro+definitions+with+a+semicolon#:~:text=Macros%20are%20frequently%20used%20to,not%20conclude%20with%20a%20semicolon.

File: utils/test/test_framework.h [hunk 2]
```diff
@@ -1,9 +1,35 @@
...
 #define TEST_FRAMEWORK_H_
 
+#include <stdio.h>
+
 #define SUCCESS 0
 #define FAIL -1
-#define ASSERT(expr) if(!(expr)) return FAIL
-#define ASSERT_IS_NONE(var) if(var != NULL) return FAIL
+
+#ifdef TEST_NO_PRINT_
```
Comment (bama4) [inline utils/test/test_framework.h]: This could be `#ifndef VERBOSE`

File: utils/test/test_framework.h [hunk 3]
```diff
@@ -1,9 +1,35 @@
...
+
+#define ASSERT(expr)                                                           \
+    if (!(expr))                                                               \
+        return FAIL;
+
+#define ASSERT_IS_NONE(var)                                                    \
+    if (var != NULL)                                                           \
+        return FAIL;
+
+#else /*TEST_NO_PRINT*/
```
Comment (bama4) [inline utils/test/test_framework.h]: The section could be run in the case where `VERBOSE` is defined

File: utils/test/test_framework.h [hunk 4]
```diff
@@ -1,9 +1,35 @@
...
+
+#ifdef TEST_NO_PRINT_
+
+#define ASSERT(expr)                                                           \
+    if (!(expr))                                                               \
+        return FAIL;
+
+#define ASSERT_IS_NONE(var)                                                    \
+    if (var != NULL)                                                           \
+        return FAIL;
```
Comment (bama4) [inline utils/test/test_framework.h]: no semicolon

File: utils/test/test_framework.h [hunk 5]
```diff
@@ -1,9 +1,35 @@
...
+
+#define ASSERT(expr)                                                           \
+    if (!(expr)) {                                                             \
+        printf("Assert failed on line %d\n", __LINE__);                        \
+        return FAIL;                                                           \
+    }
+
+#define ASSERT_IS_NONE(var)                                                    \
+    if (var != NULL) {                                                         \
+        printf("Assert failed on line %d\n", __LINE__);                        \
```
Comment (bama4) [inline utils/test/test_framework.h]: Could change to `Assert is null failed on line %d\n`

File: utils/test/test_framework.h [hunk 6]
```diff
@@ -1,9 +1,35 @@
...
+
+#else /*TEST_NO_PRINT*/
+
+#define ASSERT(expr)                                                           \
+    if (!(expr)) {                                                             \
+        printf("Assert failed on line %d\n", __LINE__);                        \
+        return FAIL;                                                           \
+    }
+
+#define ASSERT_IS_NONE(var)                                                    \
```
Comment (bama4) [inline utils/test/test_framework.h]: Could add an `ASSERT_NOT_NULL`

File: utils/test/test_framework.h (hunk 5 above)
Comment (garfr) [reply utils/test/test_framework.h]: Is "ASSERT_IS_NULL failed on line %d\n" alright?  It seems more clear to me.

**Pull request merged**